src/orange/detail/OrientedBoundingZone.cc

```diff
@@ -0,0 +1,190 @@
+        [&](real_type min_dist, auto ax) {
+            return std::min(min_dist,
+                            outer_hw[to_int(ax)] - trans_pos[to_int(ax)]);
+        });
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the safety distance for any position outside the inner box.
+ *
+ * There are two cases:
+ *
+ * Case 1: the point is between the inner and outer boxes, resulting in a
+ * safety distance of zero.
+ *
+ * Case 2: the point is outside both the inner and outer boxes, in which case
+ * the safety distance is the minimum distance from the given point to any
+ * point on the inner box. This can be calculated as:
+ *
+ * d = sqrt(max(0, px - hx)^2 + max(0, py - hy)^2 + max(0, pz - hz)^2)
+ *
+ * for a point in quadrant zero at (px, py, pz) and a box with half-widths (hx,
+ * hy, hz).
+ */
+real_type OrientedBoundingZone::safety_distance_outside(Real3 pos)
+{
+    CELER_EXPECT(!this->is_inside_inner(pos));
+
+    auto trans_pos = this->quadrant_zero(this->translate(pos));
+    auto outer_hw = storage_->half_widths[outer_hw_id_];
+
+    if (this->is_inside(trans_pos, outer_hw))
+    {
+        return 0.;
+    }
+
+    auto inner_hw = storage_->half_widths[inner_hw_id_];
+    Real3 maxes;
+    std::transform(range(Axis::size_).begin(),
+                   range(Axis::size_).end(),
+                   maxes.begin(),
+                   [&](auto ax) {
+                       return std::max(
+                           real_type{0},
+                           trans_pos[to_int(ax)] - inner_hw[to_int(ax)]);
+                   });
+
+    return std::sqrt(
+        std::inner_product(maxes.begin(), maxes.end(), maxes.begin(), 0.0));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine if a position is inside the inner box.
+ */
+bool OrientedBoundingZone::is_inside_inner(Real3 const& pos)
+{
+    return this->is_inside(this->translate(pos),
+                           storage_->half_widths[inner_hw_id_]);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine if a position is inside the outer box.
+ */
+bool OrientedBoundingZone::is_inside_outer(Real3 const& pos)
+{
+    return this->is_inside(this->translate(pos),
+                           storage_->half_widths[outer_hw_id_]);
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Translate a position to the OBZ coordinate system.
+ */
+Real3 OrientedBoundingZone::translate(Real3 const& pos)
+{
+    Real3 trans_pos;
+    TransformVisitor apply_transform(storage_->transforms, storage_->reals);
+    auto transform_down
+        = [&pos, &trans_pos](auto&& t) { trans_pos = t.transform_down(pos); };
+
+    apply_transform(transform_down, transform_id_);
+    return trans_pos;
```

Comment:
```suggestion
    TransformVisitor apply_transform(storage_->transforms, storage_->reals);
    auto transform_down
        = [&pos, &trans_pos](auto&& t) { return t.transform_down(pos); };
    return apply_transform(transform_down, transform_id_);
```
It might also be good to make a templated helper member function `auto apply_transform(F&&)` that locally creates the visitor so you can refactor the backend without changing the entire code. We don't need to keep the transform visitor around since we're only ever doing one transform per function call.

Reply:
Discussed in-person. This is already a helpful function, so the TransformVisitor appears only once.